### PR TITLE
Handle Very Large Collections In ZStream.fromIterable

### PR DESCRIPTION
**Pull request by adamgfraser**
`ZStream.fromIterable` is a little tricky because most of the time the collection is a `Chunk` that we can just put into a stream but the collection could potentially be infinite or even if it is finite it could use a "sparse" representation like a `Range` where converting it to a "dense" representation of its values could exceed system resources. We can continue just putting `Chunk` values into the stream to optimize the fast path while otherwise going through `fromIterator` for safety.

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
Thank you for picking up so quickly! I was rewriting some akka-stream code when the issue came up, it was something like `Source(largecollection).map(allocateLargeObject)` and the zio-stream version blew up. That made me realise how important understanding the underlying chunk-based nature is to manipulating streams.

Then I wondered if I should use `rechunk` but after playing with that thought I wasn't really satisfied. It is a bit of an abstraction leak, but also I don't really care if alll chunks are the same size, what I think I would be up for is to set an upper bound. After looking at the implementation of `Chunk.flatMap`, it seems like `ZStream.fromIterable(largeChunk).mapConcat(createLargeChunk)` would still end up with a stream containing a single chunk? If that is the case, then I think this may be a sign the parameter is not very compositional.

In summary, I was wondering if a regional combinator like `withParallelism` that would set the max chunk size would be a good idea?

**Comment by adamgfraser** [comment]
@regiskuckaertz Yes we have some ideas on this that I think will come as part of ZIO 2.1

**Review by jdegoes — APPROVED** [review]
